Generate core API docs from TSDoc comments

rudolf opened the pull request:
## Summary

Fixes #31717 

Uses `api-extractor` and `api-documenter` to generate documentation for the Kibana core API from TSDoc comments in the source code. This does not integrate the generated docs into the kibana documentation workflow and won't get published to the website. Instead, these markdown files are meant to be viewed in your editor or on Github. [Example of viewing docs in Github](https://github.com/elastic/kibana/blob/716fa7549c2ae0ade7ff2ba52985fa06e91ce161/docs/development/core/api/kibana.logger.md)

Documentation can be generated using `node scripts/check_core_api_changes --docs`

Todo:
 - [x] Take a first stab at fleshing out docs like parameter and method descriptions so they're actually useful. Also add a few missing documentation tags that currently causes api-extractor to fail with a non-zero exit code.
 - [x] Add test that checks for core api changes by comparing the API signature with `common/core_api_review/kibana.api.ts`. Running `node scripts/check_core_api_changes --accept` will update the review file and documentation and has to be commited to pass the test.

## Details
To generate the documentation the following steps are executed:
1. `npm run build:types` (pre-existing functionality in our build system)
   Uses `src/type_exports.ts` to create `target/types/type_exports.d.ts`. These types are exported when someone does `import {...} from 'kibana'` (`package.json` has a "types" property pointing to `./kibana.d.ts` which re-exports `target/types/type_exports.d.ts`)
2. `api-extractor` uses `target/types/type_exports.d.ts` to determine the surface area of the public API and generates `./build/kibana.api.json`. 
   This is like an AST of the API and it's TSDoc comments used in (3) by `api-documenter` to render markdown documentation. 
   It also generates an API review file `build/kibana.api.ts` and compares this file to `src/core/public/kibana.api.md`. A difference in the type signatures means the public API signature has been changed which will print out a warning. A developer has to "accept" the change by running `node scripts/check_core_api_changes --accept` and then committing the updated API review file and documentation.
3. `api-documenter` creates documentation by "rendering" the contents of `./build/{public/server}/kibana.api.json` as markdown files which we commit to the repository.

Since api-exporter doesn't support `import * as Module from __`, a technique we use to prevent type collisions in `src/core/index.ts` we treat the public and server api's as two different packages and run api-extractor twice. The biggest drawback is that api-extractor uses the name field in `package.json` as for api review files and documentation.

### Checklist

Use ~~strikethroughs~~ to remove checklist items you don't feel are applicable to this PR.

- [ ] ~~This was checked for cross-browser compatibility, [including a check against IE11](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#cross-browser-compatibility)~~
- [ ] ~~Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/master/packages/kbn-i18n/README.md)~~
- [ ] ~~[Documentation](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#writing-documentation) was added for features that require explanation or tutorials~~
- [ ] ~~[Unit or functional tests](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#cross-browser-compatibility) were updated or added to match the most common scenarios~~
- [ ] ~~This was checked for [keyboard-only and screenreader accessibility](https://developer.mozilla.org/en-US/docs/Learn/Tools_and_testing/Cross_browser_testing/Accessibility#Accessibility_testing_checklist)~~

### For maintainers

- [ ] ~~This was checked for breaking API changes and was [labeled appropriately](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#release-notes-process)~~
- [ ] ~~This includes a feature addition or change that requires a release note and was [labeled appropriately](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#release-notes-process)~~

rudolf committed:
Generate core API docs from TSDoc comments

Uses api-extractor and api-documenter to generate documentation for
the Kibana core API from TSDoc comments in the source code.

Documentation can be generated using `npm run docs:api`.

I used --no-verify to ignore the following pre-commit hook errors:
1. Filenames MUST use snake_case - api-extractor.json
   It's possible to specify a different config file, but I prefer to keep the "standard" config file name.
2. UNHANDLED ERROR: Unable to find tsconfig.json file selecting "common/core_api_review/kibana.api.ts". Ensure one exists and it is listed in "src/dev/typescript/projects.ts"
   This is not a source file, so safe to ignore.

elasticmachine commented:
Pinging @elastic/kibana-platform

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/2818/)

rudolf committed:
Flesh out API docs a little bit

rudolf committed:
Ignore snake_case check for api-extractor.json

rudolf committed:
Ignore api-extractor's review file from pre-commit check

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/2941/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/2947/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/2952/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/3121/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/3140/)

rudolf committed:
Try to fix build failing by using masters yarn.lock

rudolf committed:
I'm being stupid

rudolf committed:
Found a better home for ignoring common/core_api_review/kibana.api.ts

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/3150/)

rudolf committed:
Node script for detecting core API changes

I initially wanted to include this as a precommit hook, but it takes
quite long to execute (~12s) so might be better suited as a test or
as part of the release process.

The script currently fails because api-extractor uses an older version
of typescript.

skaapgif commented:
Checking for core API changes takes quite long, which can get annoying in a pre-commit hook, so I think it's better suited as a test.

`api-exporter` doesn't yet use the latest typescript compiler introduced in (#32215) so it returns a non-zero exit code which means we can't enforce it as a test yet.

I suggest that we don't block on `api-exporter` and merge this so long, manually keeping docs updated. Once it supports the latest typescript we can update `api-exporter` and automate the detection of api changes as a test.

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/3163/)

spalger commented on src/dev/typescript/get_ts_project_for_absolute_path.ts:
```diff
@@ -31,8 +31,9 @@ import { PROJECTS } from './projects';
+export function getTsProjectForAbsolutePath(path: string): Project {
   const relPath = relative(REPO_ROOT, path);
+
   const projects = PROJECTS.filter(p => p.isAbsolutePathSelected(path));
```
Personally, I'd prefer if we defined this filter function as a type guard so that TS can infer the return value, and so that `projects` has the right type inside the function.

skaapgif replied:
I'm not sure how you would use a type guard here since `projects` can only ever be one type (`Projects[]`). TS correctly infers this as being of type `Project` since the `filter` function is typed.

I haven't given this much thought, but I think I prefer exported functions explicitly defining their return types instead of inferring it from returned values only. If you change the implementation of the function, you will immediately see a warning that you've broken the type contract as you're working inside of that file. If the type is inferred, consumers of the function would throw type errors, but you probably won't notice this as quickly.

joshdover replied:
We don't have a defined convention here. I think if that's one we want to adopt, we should introduce a PR with a linter rule for it, allowing for a team discussion. IMO anything that is not enforced by a lint rule is free game. This philosophy encourages us to have one discussion about it, make a decision, and commit to it rather than having the same argument over and over.

spalger replied:
Yeah, I was wrong in my original comment, and I agree @joshdover.

restrry commented on src/dev/tslint/pick_files_to_lint.ts:
```diff
@@ -18,9 +18,17 @@
+    file =>
+      file.isTypescript() &&
+      !file.isFixture() &&
+      IGNORED_PATHS.some(glob => minimatch(file.getRelativePath(), glob, { dot: true }))
```
shouldn't we use inversion here? how I can test this script?

skaapgif replied:
Good catch! I've pushed a commit to fix it. This script runs as a precommit hook, `files` is an array of staged files, so to test you need to stage some changes with `git add` and then try to commit it.

rudolf committed:
Fix tslint precommit hook ignore condition

skaapgif commented on package.json:
```diff
@@ -48,6 +48,7 @@
     "test:ui:runner": "node scripts/functional_test_runner",
     "test:server": "grunt test:server",
     "test:coverage": "grunt test:coverage",
+    "docs:api": "npm run build:types && api-extractor run && api-documenter markdown -i ./build -o ./docs/development/core/api",
```
Yeah, api-extractor looks for `typings` and we use `types` in package.json. Both are valid according to the typescript website. I've submitted https://github.com/Microsoft/web-build-tools/pull/1115 which will fix it, but it's probably a good idea to specify both values in our package.json until it's been published upstream.

skaapgif replied:
The upstream change would have written `tsdoc-metadata.json` to the toplevel kibana directory, so I've pushed a commit now that specifies the `tsdocMetadata` location in package.json so that it gets written to `./build`.

rudolf committed:
Write tsdoc-metadata.json into ./build

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/3416/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/3422/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/3493/)

rudolf committed:
Add LogMeta and ElasticSearch to exported types & docs

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/3589/)

rudolf committed:
Merge branch 'master' into rm-core-api-docs

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/3652/)

joshdover commented on common/core_api_review/kibana.api.ts:
```diff
@@ -0,0 +1,72 @@
+// @public
```
What is this core_api_review directory exactly? Just any "public" exported type? How does api-extractor/api-documentor use it? Would it make sense to just gitignore this?

skaapgif replied:
I'll expand a bit on how this works in the PR description, it's probably valuable context for anyone reviewing this.

rudolf committed:
Suppress logging when running api-extractor from script

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/3757/)

rudolf committed:
Improve check_core_api_changes script and run as test

rudolf committed:
Inline api-extractor.json config

skaapgif commented on src/dev/run_check_core_api_changes.js:
```diff
@@ -0,0 +1,40 @@
+import { Extractor } from '@microsoft/api-extractor';
+import { run } from './run';
+
+const config = require('../../api-extractor.json');
```
I started using the api-extractor command which requires a config file, but now that it's all happening within a script I've inlined the configuration.

skaapgif commented on package.json:
```diff
@@ -48,6 +48,7 @@
     "test:ui:runner": "node scripts/functional_test_runner",
     "test:server": "grunt test:server",
     "test:coverage": "grunt test:coverage",
+    "docs:api": "npm run build:types && api-extractor run && api-documenter markdown -i ./build -o ./docs/development/core/api",
```
removed this from package.json as it's now replaced with a grunt task run as part of "test".

rudolf committed:
Merge branch 'master' into rm-core-api-docs

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/3887/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/3894/)

restrry commented on src/dev/run_check_core_api_changes.ts:
```diff
@@ -0,0 +1,187 @@
+  const extractor = new Extractor(apiExtractorConfig, options);
+  extractor.processProject();
+
+  errors.forEach(msg => log.error(msg));
```
according to https://github.com/Microsoft/web-build-tools/blob/master/apps/api-extractor/src/api/ILogger.ts any error or warning will break prod build. should we do `process.exit(1)` on any such message, not just `apiChanged`?

skaapgif replied:
I've made it more permissible than their default so that warnings about missing documentation tags or incorrect TSDoc comments doesn't break the build. 

But perhaps it's better to start with a strict implementation and only relax it if we realise it is introducing annoying problems.

restrry commented on src/dev/run_check_core_api_changes.ts:
```diff
@@ -0,0 +1,187 @@
+
+  errors.forEach(msg => log.error(msg));
+  warnings
+    .filter(msg => !msg.startsWith('You have changed the public API signature for this project.'))
```
seems we use private data (message in this case) as a contract, which can be error-prone to changes when we update library version. can we just check error/warning presence?

skaapgif replied:
Yes this is a very brittle integration and not a nice way to consume a library. I've created https://github.com/Microsoft/web-build-tools/issues/1133 and they're hoping to address it before the release of `api-extractor` v7 (we're using the beta).

rudolf committed:
Fix check_core_api_changes --help flag

rudolf committed:
LogMeta TSDoc comments

rudolf committed:
check_core_api_changes: fail if api-extractor produces warnings or errors

And print more useful messages to the console

skaapgif commented:
<img width="882" alt="screenshot 2019-03-07 at 17 03 42" src="https://user-images.githubusercontent.com/1281646/53970500-26986000-40fb-11e9-91e4-ee5e064d49ed.png">

@restrry I've also improved the output of the script

skaapgif commented on src/dev/run_check_core_api_changes.ts:
```diff
@@ -0,0 +1,187 @@
+        Options:
+
+          --accept    {dim Accepts all changes by updating the API Review file \`common/core_api_review/kibana.api.ts\` and documentation}
+          --help      {dim Show this message}
```
`run` intercepts the help flag and prints it's own help... as it's quite a thin wrapper and other scripts depend on this behaviour I removed it.

skaapgif commented on src/dev/run_check_core_api_changes.ts:
```diff
@@ -0,0 +1,187 @@
+import getopts from 'getopts';
+import { run } from './run';
+
+const apiExtractorConfig: IExtractorConfig = {
```
Since it's used for generating the documentation we want others to use, I think it might be better to be explicit about which API's are public, beta or private.

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/4096/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/4113/)

mshustov reviewed (approved):
LGTM

rudolf committed:
Merge branch 'master' into rm-core-api-docs

rudolf committed:
Move ignored ts files list into dev/file

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/4550/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/4868/)

rudolf committed:
Merge branch 'master' into rm-core-api-docs

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/5012/)

spalger commented on common/core_api_review/kibana.api.ts:
```diff
@@ -0,0 +1,72 @@
+// @public
```
I'm sorta conflicted on the choice to put this in the "common" directory at the root of the repository. I'm not sure why this directory exists in the first place, but could you help me understand why this placement makes sense?

I'm conflicted because I can't really think of a _better_ place to put it, though I feel like we might be able to come up with something a little more self explanatory than "common".

skaapgif replied:
Yeah I don't really like "common" either, it doesn't mean anything to me or describe to me what's inside of this folder. The `common` directory is the convention used by api-generator and Microsoft uses that convention in most of their projects. The graphql introspection files seem to follow the same convention, so I thought I'll stick to that in the absence of any better ideas.

We don't want to put this in `/src` cause it is a generated file, not a source file. We don't want it in `/build` cause it needs to get commited so that we can compare changes to the api signature. We could potentially put this is `/dev` or create a more descriptive top-level directory like `api_signature` which could also be the home for the GraphQL introspection files.

epixa replied:
If it helps, we have one precedent in the repo today for committing a generated file, and that is under a `dist` directory at the root of that package: https://github.com/elastic/kibana/tree/master/packages/kbn-pm/dist

skaapgif replied:
api-extractor now creates a .md file so it's safe to put the API review file with the source code, so I've moved the review file into the /public /server directories.

spalger commented on common/core_api_review/kibana.api.ts:
```diff
@@ -0,0 +1,72 @@
+// @public
+declare class ClusterClient {
+    // (undocumented)
+    constructor(config: ElasticsearchClientConfig, log: Logger);
```
Is there any way to ensure that all the types referenced in the config are included in the API? Even though `ElasticsearchClientConfig` isn't exported it's still a part of the "public api" right? Changes to that class should be considered a change to the Public API I would think.

skaapgif replied:
I'll follow-up upstream, but I agree this seems like a bug. I also noticed that generating markdown documentation doesn't create a paragraph for a class' constructor (the yaml generator does). I'll keep tabs on or help fix the issues upstream, but don't want to block this PR on that.

spalger commented on docs/development/core/api/kibana.clusterclient.md:
```diff
@@ -0,0 +1,25 @@
+[Home](./index) &gt; [kibana](./kibana.md) &gt; [ClusterClient](./kibana.clusterclient.md)
```
All of these docs mention an index file that doesn't exist, can we generate one?

skaapgif replied:
We could create one, but not sure what we would put inside it? Perhaps we can create a symlink that points to `./kibana`, then at least you don't click through to a page that doesn't exist.

spalger replied:
I was actually kinda hoping it would have a table of contents listing the existing docs or something like that. I think it'd be nice to see from a high level what is/isn't documented.

spalger reviewed (commented):
Looking great so far

skaapgif commented on src/dev/run_check_core_api_changes.ts:
```diff
@@ -0,0 +1,201 @@
+
+  const { apiChanged, warnings, errors } = runApiExtractor(opts.accept);
+
+  if (apiChanged) {
```
The reason behind the late panic was that I wanted to exit with a non-zero exit when there's warnings so that tests fail, but apart from tests, it's more useful to generate artifacts as far as possible. For instance, forgetting to add an `@public` or `@private` tag should fail tests, but it's still nice to see the documentation output while you're developing so you can see how your changes get "rendered".

Having said that, I'll take a look at this to see if return statements could simplify the if..else logic.

rudolf committed:
Merge branch 'master' into rm-core-api-docs

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/5979/)

rudolf committed:
Add back build:types since api-exporter cannot operate on source files

rudolf committed:
Upgrade api-exporter/documenter

rudolf committed:
api-extractor: independantly analyze core/public and core/server

Becasue of https://github.com/Microsoft/web-build-tools/issues/1029
api-extractor can't use core/index.ts as a single entry point for
analyzing the public and server API's as isolated namespaces.

Instead we analyze these projects separately. This introduces other
problems like the api review files and documentation always being
called "kibana." from the package.json filename.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/6348/)

rudolf committed:
Build types as part of build task

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/6455/)

rudolf committed:
Include types in typescript browser compilation

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/6515/)

rudolf committed:
Force inclusion of core/public for building types

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/6525/)

rudolf committed:
Merge remote-tracking branch 'upstream/master' into rm-core-api-docs

rudolf committed:
Fix api review filename in api-exporter errors

rudolf committed:
Update docs and API review files

rudolf committed:
Fix api-extractor warnings

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/6540/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/6547/)

rudolf committed:
Merge branch 'master' into rm-core-api-docs

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/6777/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/7161/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/7204/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/7263/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/7341/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/7347/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/7350/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/7353/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/7364/)

mistic committed:
chore(NA): merge and solve conflicts with last master.

rudolf committed:
Remove ts file ignored list since it's no longer necessary

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/7438/)

rudolf committed:
Rename exported api package name

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/7445/)

restrry commented on src/core/server/kibana.api.md:
````diff
@@ -0,0 +1,110 @@
+
+```ts
+
+import { ConfigOptions } from 'elasticsearch';
````
interesting, why extractor keeps unused imports?

skaapgif commented on src/dev/run_check_core_api_changes.ts:
```diff
@@ -0,0 +1,245 @@
+  await runBuildTypes().catch(e => {
+    if (e) {
+      log.error(e);
+      return false;
```
If one of the projects' typescript compiler fails to compile the types, I don't want it to exit both processes. Instead I exit the process with `exitCode = 1` on line 243.

restrry replied:
but how? it's not `return await runBuildTypes()`

skaapgif replied:
Ah I see now, yes I should fix that.

monfera reviewed (approved):
LGTM, I don't see much in the way of Canvas impact, @clintandrewhall maybe worth a look?

skaapgif commented on src/dev/run_check_core_api_changes.ts:
```diff
@@ -0,0 +1,245 @@
+  );
+};
+
+const isApiChangedWarning = (warning: string) => {
```
good idea

skaapgif commented on src/dev/run_check_core_api_changes.ts:
```diff
@@ -0,0 +1,245 @@
+  }
+
+  // If any errors or warnings occured, exit with an error
+  if (errors.length > 0 || warnings.length > 0) {
```
api-extractor gives warnings for problems like "Warning: (ae-internal-missing-underscore) The name bootstrap should be prefixed with an underscore because the declaration is marked as "@internal"" It's possible to silence some of the rules in the latest beta releases, but I found it very helpful to be able to inspect the generated documentation, even if there were warnings present that would break the build.

Once api-extractor v7 is released we can clean this up a lot.

skaapgif commented:
@monfera Thanks for the review. This shouldn't impact any other teams, but due to a large merge commit, other teams were incorrectly tagged as code owners.

skaapgif commented on src/dev/run_check_core_api_changes.ts:
```diff
@@ -0,0 +1,245 @@
+};
+
+const renameExtractedApiPackageName = async (folder: string) => {
+  const json = JSON.parse(await fs.readFileSync(`build/${folder}/kibana.api.json`).toString());
```
:+1:

skaapgif commented on docs/development/core/server/kibana-plugin-server.plugininitializercontext.config.md:
```diff
@@ -0,0 +1,12 @@
+[Home](./index) &gt; [kibana-plugin-server](./kibana-plugin-server.md) &gt; [PluginInitializerContext](./kibana-plugin-server.plugininitializercontext.md) &gt; [config](./kibana-plugin-server.plugininitializercontext.config.md)
```
yeah, I can submit that upstream to api-documenter.

skaapgif commented on src/dev/build/tasks/typecheck_typescript_task.js:
```diff
@@ -25,48 +25,87 @@ export const TypecheckTypescriptTask = {
+      ...typesProjectBuild.config,
+      files: [
+        ...(typesProjectBuild.config.files || []),
+        'src/core/public/index.ts'
```
There are two problems here, we mutate configuration files and these configuration files inherit from each other. So the exclude rule in `defaultProjectBuild` excludes `src/core/server/index.ts` and then we have to mutate the types config file in the build folder to include that file back.

I think we need to clean up the way we compile/check typescript. I first had this as an inline option, but it was actually really useful to have all the options in a configuration file for debugging (even though it's confusing that it gets mutated after this script runs).

rudolf committed:
Review comments

rudolf committed:
Export other missing types

skaapgif commented on docs/development/core/server/kibana-plugin-server.clusterclient.asscoped.md:
```diff
@@ -0,0 +1,24 @@
+
+|  Parameter | Type | Description |
+|  --- | --- | --- |
+|  req | `{`<p/>`        headers?: Headers;`<p/>`    }` | Request the `ScopedClusterClient` instance will be scoped to. |
```
I've exported these types, but api-documenter doesn't automatically link them up.

restrry replied:
this is bummer 😞 ... probably we should file some issues in their repo and track progress. 
not a blocker for the current PR

skaapgif commented on docs/development/core/server/kibana-plugin-server.clusterclient.md:
```diff
@@ -0,0 +1,25 @@
+
+## Properties
+
+|  Property | Modifiers | Type | Description |
```
This seems to be things like `@sealed`, `@virtual` & `@override`.

rudolf committed:
Upgrade api-documenter to latest beta

rudolf committed:
Export more missing types

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/7784/)

rudolf committed:
Fix warnings and add api-exporter to Jenkins tests

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/7802/)

restrry commented on docs/development/core/server/kibana-plugin-server.pluginsetupcontext.md:
```diff
@@ -14,5 +14,5 @@ export interface PluginSetupContext
 |  Property | Type | Description |
 |  --- | --- | --- |
-|  [elasticsearch](./kibana-plugin-server.pluginsetupcontext.elasticsearch.md) | `{`<p/>`        adminClient$: Observable<ClusterClient>;`<p/>`        dataClient$: Observable<ClusterClient>;`<p/>`    }` |  |
+|  [elasticsearch](./kibana-plugin-server.pluginsetupcontext.elasticsearch.md) | <code>{`<p/>`        adminClient$: Observable&lt;ClusterClient&gt;;`<p/>`        dataClient$: Observable&lt;ClusterClient&gt;;`<p/>`    }</code> |  |
```
ooooh. they fix one thing and break another. let's don't block this PR
![2019-04-02_22-07-22](https://user-images.githubusercontent.com/3198181/55432689-d98d9980-5593-11e9-892a-0fc112dd949b.png)

mshustov reviewed (commented):
I think there are several issues in documenter, so it's not a blocker for the current PR, but nice to fix later
- cross links in types. when a typeA is mentioned in `TypeB.md`, there should be a link to `typeA.md` page
- weird alignments in `typescript` code blocks
- escaped line break for object types `{<p/> adminClient$: Observable<ClusterClient>;<p/> dataClient$: Observable<ClusterClient>;<p/> }`
- warning comment for autogenerated documentation page

only problem in the code
https://github.com/elastic/kibana/pull/32148/files/51b8bc1f09219075eeb7209e5a93c3db6c91474e#r270865425

rudolf committed:
Merge remote-tracking branch 'upstream/master' into rm-core-api-docs

rudolf committed:
Correctly handle runBuildTypes() exceptions

rudolf committed:
Fix another swallowed exception

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/7967/)

rudolf committed:
Fix api-extractor warnings after master merge

skaapgif commented:
After the latest merge there are still undocumented classes and missing exports. Some of them are related to #34416 . I want to merge this to plug the leaky bucket of new changes being committed to master which causes api-extractor to fail. After #34416 we can do a cleanup of the exported types and documentation.

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/7976/)

skaapgif commented:
> * cross links in types. when a typeA is mentioned in `TypeB.md`, there should be a link to `typeA.md` page

I've created https://github.com/Microsoft/web-build-tools/issues/1237
> * weird alignments in `typescript` code blocks

4-space indentation is the most widely supported markdown standard, so I believe that's why api-documenter uses it.
> * escaped line break for object types `{<p/> adminClient$: Observable<ClusterClient>;<p/> dataClient$: Observable<ClusterClient>;<p/> }`

Upstream PR https://github.com/Microsoft/web-build-tools/pull/1234
> * warning comment for autogenerated documentation page

Fixed in https://github.com/elastic/kibana/pull/34817
